core/sparse_array.cpp

```diff
@@ -38,14 +38,32 @@ SparseArray::SparseArray(unsigned int p_nrows, unsigned int p_ncols)
 	nrows_set_ = 0;
 	nnz_ = 0;
 	nnz_capacity_ = 1024;
-	
 	row_offsets_ = (uint32_t *)malloc((nrows_ + 1) * sizeof(uint32_t));
 	columns_ = (uint32_t *)malloc(nnz_capacity_ * sizeof(uint32_t));
 	distances_ = (sa_distance_t *)malloc(nnz_capacity_ * sizeof(sa_distance_t));
 	strengths_ = (sa_strength_t *)malloc(nnz_capacity_ * sizeof(sa_strength_t));
 	
 	row_offsets_[nrows_set_] = 0;
 	finished_ = false;
+
+    //Allocate space for 2-D vectors
+    //Columns and Distances are now ragged 2-D vectors with each row corresponding to one focal individual. This negates the need for row offsets.
+
+    initial_width=100;       //Some random initial width for each row. Can be changed later.
+    columns = (uint32_t **)malloc((nrows_) * sizeof(uint32_t *));
+    distances = (sa_distance_t **)malloc((nrows_) * sizeof(sa_distance_t *));
+    nnz = (uint32_t *)malloc((nrows_) * sizeof(uint32_t));
+    nnz_capacity = (uint32_t *)malloc((nrows_) * sizeof(uint32_t));
+
+    for(int i=0;i<nrows_;i++)
```

Comment:
Much better than looping to zero out these vectors would be to use calloc() instead of malloc() above.  That lets the kernel zero out the memory for us, which can be *way* faster than doing it ourselves.  Whenever you write a loop to zero out memory, think about whether you could use calloc() (for a new allocation) or bzero (for an existing allocation).